src/dlm.c

```diff
@@ -0,0 +1,926 @@
+
+   Our approach is to queue all the tiles in our own queue, and return immediately
+   while letting a download thread feed curl the most recently added tiles first.
+
+   This way, if you scroll past an area, the tiles will all be queued for checking
+   but we will always start downloading tiles in the current view (ie most
+   recently added) first - without cancelling any tiles already started.
+
+
+   Mini HowTo:
+
+   in your drawing function, before reading tiles, do this:
+
+      DLM_queue_tile(...) each map tile you want
+
+      call DLM_do_transfers() - this only does anything if the DLM was compiled
+                                with threading disabled
+
+      optionally call DLM_wait_done() if you want to wait for all downloads to finish
+
+      Draw your maps using the tiles available in your cache.
+
+   When each download completes, the global request_new_image is incremented
+   which triggers a redraw from the top level.  It is safe to repeat this whole
+   sequence at each redraw as tiles are only added to the queue if they are
+   not in the queue already and the file does not exist, or the file is older
+   than 7 days.
+
+*/
+
+// Enable use of a separate download thread in the background
+#define DLM_QUEUE_THREADED
+
+// Enable using curl_multi_* to download more than one tile at a time
+// Define this to use the code, set the number to limit the
+// simultaneous transfers
+// WARNING: defining this less than 1 will allow starting an unlimited number
+// of transfers. You probably shouldn't do that!
+// Also, don't set this too high or slow internet connections will have to
+// wait too long before starting to download tiles in the current view
+// when the user scrolls around a bit
+#define USE_CURL_MULTI 8
+
+
+#define DLM_Q_STOP	0
+#define DLM_Q_STARTING	1
+#define DLM_Q_RUN	5
+#define DLM_Q_IDLE	8
+#define DLM_Q_QUIT	9
+
+#define MAX_DESCLEN	40
+
+struct DLM_queue_entry {
+    struct DLM_queue_entry *next;
+    struct DLM_queue_entry *prev;
+
+    // These are only used by DLM_queue_tile for
+    // checking if a tile is already queued
+    // osm_zl is also used as a type flag (>=0 for tile, <0 for file)
+    unsigned long	  x;
+    unsigned long	  y;
+    int			  osm_zl;
+
+    int			  state;
+    xastir_mutex	  lock;
+    char		  fileName[MAX_FILENAME];
+    char		  desc[MAX_DESCLEN];
+    char		  *tempName;
+    char		  *url;
+
+#ifdef HAVE_LIBCURL
+    FILE		  *stream;
+#ifdef USE_CURL_MULTI
+    char		  *curlErrBuf;
+    CURL		  *curlSession;
+#endif
+#endif
+};
+
+
+/* DLM_queue_lock is held whenever reading or changing DLM_queue_entry or
+ * the linkages of the items in the queue.  Each item also has a lock
+ * for changing data in that item.
+ */
+xastir_mutex DLM_queue_lock = { .threadID=0, .lock=PTHREAD_MUTEX_INITIALIZER };
+pthread_t DLM_queue_thread;
+
+volatile int DLM_queue_progress_flag=0;
+volatile int DLM_queue_state=DLM_Q_STOP;
```

Comment:
Should DLM_queue_state be protected by a mutex to ensure correct memory visibility? I don't believe volatile is enough.

Reply:
Are you ok if I merge this while we wait on the answer? I'm good with the rest of the code and think this would be a step forward. If the variable needs protection one of us can add that.

Reply:
I am OK with that, although it might be good to hold on a new release until we figure it out.

Reply:
I'm confused as to how DLM_queue_state should be working. I could see it protected by a mutex and being an INT (for perhaps a count of download threads running), but I don't see how having multiple threads able to write specific values into it will work, mutex or not. It's a single variable but multiple threads can write STOP/RUN/QUIT/IDLE/STARTING values into it at any time. Perhaps make it a count variable with defined ALL-STOP value (-1) or something? It's also easily possible that I misunderstand what its purpose is.

Reply:
Just to clarify what my concern is: I've been bitten in the past where only one thread is modifying a variable and the reading thread does not see the update. The variable was marked volatile. A mutex is needed to ensure changes are visible to other threads. Having the mutex inserts a memory barrier.

See https://en.wikipedia.org/wiki/Memory_barrier#Example for a bit of an explanation.

Reply:
I can add in the mutexes if desired.

Reply:
Feel free. Usual fork / modify(on a branch) / pull-request. I'd have to look deeper at the code to decide for sure that it's needed, but if one thread is writing and at least one other is reading - we need it. Thanks!